rad/chain.rad

```diff
@@ -38,52 +55,91 @@
-    (def ch (ref chain))
-    (def inputs (ref (list)))
-    (eval-definer
-     (fn [e]
-       (cond
-        (eq? e :quit) (do (print! "Leaving remote-chain.")
-                          (eval-definer old-eval))
-        (eq? e :send) (do (print! "Sending inputs to remote chain.")
-                          (map (fn [i] (send! chain-id i)) (reverse (read-ref inputs)))
-                          (print! "Done! Leaving remote-chain.")
-                          (eval-definer old-eval))
-        :else (do (def x (eval-in-chain e (read-ref ch)))
-                  (write-ref ch (view (@ :chain) x))
-                  (modify-ref inputs (fn [is] (cons e is)))
-                  (view (@ :result) x)))))))
-
-(document 'enter-remote-chain
-          '(("eval-definer" function) ("chain-id" string))
-          "Syncs up to a remote chain, and allows experimentation inputs. There are two special inputs: ':quit' will leave the chain, ':send' will send all the accumulated inputs to the remote chain, one by one, and then leave the chain. Uses the parameter 'eval-definer' for eval redefinition, this should be a function which can be used to redefine 'eval': it should set eval to it's input.")
-
-;; Setup an eval definer
-(def boom
-  (quote
-   (do (def eval-ref (ref eval))
-       (def eval-definer (fn [f] (write-ref eval-ref f)))
-       (def eval (fn [e] ((read-ref eval-ref) e))))))
+;; update-chain
+
+(def update-chain
+  (fn [chain]
+    (def next (ref chain))
+    (update-chain-with chain (lookup :name chain) (fn [x] (modify-ref next (fn [r] (eval-in-chain x r)))) print!)))
+
+(document 'update-chain
+		  '(("chain" chain))
+		  "Return a version of 'chain' with new inputs and correspondingly updated state")
+
+;; add-quit
+
+(def add-quit
+  (fn [old-state new-eval]
+    (fn [expr env]
+      (if (eq? expr :quit)
+          (list :ok old-state)
+          (new-eval expr env)))))
+
+;; enter-chain
+
+;; TODO move else-where
+(def simple-eval
+ (fn [expr]
+     (head (eval expr (get-current-env)))))
+
+;; store-exprs
+
+(def store-exprs
+ (fn [evalfn]
+     (fn [expr env]
+         (def env_ (over (.. (@ :env) (@ '_inputs)) (fn [x] (add-right expr x)) env))
+         (evalfn expr env_ ))))
+
+(document 'store-exprs
+          '(("eval" fn)) 
+          "Store each new evaluated expression in '_inputs'")
+
+;; add-send
+(def add-send
+  (fn [oeval]
+      (fn [expr env] 
+          (if (eq? expr :send) 
+              (do (def inpl (.. (@ :env) (@ '_inputs))) 
+                  (def chainl (.. (@ :env) (@ '_cur-chain)))
+                  (map (fn [e] (send! (view chainl env) e)) (view inpl env))
+                  (def env_ (set inpl [] env))
+                  (list :ok env_))
+              ((store-exprs oeval) expr env)))))
+
+(document 'add-send
+          '(("eval-fn" fn))
+          "Add a :send special form that sends the contents of _input to the chain _cur-chain")
+
+;; enter-chain & enter-remote-chain
+
+(def eval__ eval)
```

Comment:
Is this file part of the prelude? Or is this meant to capture "the current eval" when the file is `load!`ed?

Reply:
Yeah, so this is worth thinking a bit about. Currently it captures whatever eval happens to be in place at the moment this is run. If the only eval redefs we make are like this - in particular, if they're only adding commands, and commute, then I suppose it doesn't matter. But they won't always...

Reply:
I think it's okay for now.